include/RandBLAS/dense.hh

```diff
@@ -0,0 +1,181 @@
+#pragma once
+
+#ifndef BLAS_HH
+#include <blas.hh>
+#define BLAS_HH
+#endif
+
+#ifndef RandBLAS_STATE_HH
+#include <RandBLAS/state.hh>
+#endif
+
+#ifndef RandBLAS_DO_HH
+#define RandBLAS_DO_HH
+
+/*
+Paradigm for APIs involving structs:
+    Free-functions when there's no memory to manage
+    Member functions when there IS memory to manage, or in initializing.
+        We want to make this library hard to misuse in C++.
+    We provide APIs that we require people use to ensure that structs are
+        in a valid state. If you want to initialize the struct yourself
+        we won't stop you, but we also take no responsibility for the 
+        inevitable segfaults.
+
+TODO: have a discussion around using smart pointers for memory safety.
+    Burlen thinks we should seriously consider using smart pointers.
+*/
+
+namespace RandBLAS::dense {
+
+enum class DenseDistName : char {
+    Gaussian = 'G',         
+    Uniform = 'U',          // uniform over the interval [-1, 1].
+    Rademacher = 'R',       // uniform over {+1, -1}.
+    Haar = 'H',             // uniform over row-orthonormal or column-orthonormal matrices.
+    DisjointIntervals = 'I' // might require additional metadata.
+};
+
+struct DenseDist {
+    const DenseDistName family = DenseDistName::Gaussian;
+    const int64_t n_rows;
+    const int64_t n_cols;
+};
+
+template <typename T>
+struct DenseSkOp {
+    const DenseDist dist;
+    const RNGState state;
+    const bool own_memory = true;
+    /////////////////////////////////////////////////////////////////////
+    //
+    //      Properties specific to dense sketching operators
+    //
+    /////////////////////////////////////////////////////////////////////
+
+    T *buff = NULL;
+    bool filled = false;
+    bool persistent = true;
+    const blas::Layout layout = blas::Layout::ColMajor;
+
+    /////////////////////////////////////////////////////////////////////
+    //
+    //      Member functions must directly relate to memory management.
+    //
+    /////////////////////////////////////////////////////////////////////
+
+    //  Elementary constructor: needs an implementation
+    DenseSkOp(
+        DenseDist dist_,
+        RNGState state_,
+        T *buff_,
+        bool filled_,
+        bool persistent_,
+        blas::Layout layout_
+    );
+
+    DenseSkOp(
+        DenseDist dist,
+        uint32_t ctr_offset,
+        uint32_t key,
+        T *buff,
+        bool filled,
+        bool persistent,
+        blas::Layout layout
+    ) : DenseSkOp(dist, RNGState{ctr_offset, key}, buff, filled, persistent, layout) {};
```

Comment:
Note to self: It seemed that for SparseSkOp to work, I had to move these kinds of constructors into the .cc file, not leave in the .hh file. But this suggests leaving in the .hh file is fine.